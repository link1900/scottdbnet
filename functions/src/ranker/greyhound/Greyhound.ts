import { Entity, Column } from 'typeorm';
import { BaseModel } from '../../database/BaseModel';

@Entity()
export class Greyhound extends BaseModel {
  // @ts-ignore
  @Column() name: string;

  @Column({ type: 'uuid', nullable: true })
  sireId?: string;

  @Column({ type: 'uuid', nullable: true })
  damId?: string;

  @Column({ nullable: true })
  color?: string;

  @Column({ nullable: true })
  gender?: string;

  @Column({ nullable: true })
  dateOfBirth?: Date;
}
