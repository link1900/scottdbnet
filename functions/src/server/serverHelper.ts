import { getVariable } from '../environment/environmentHelper';
import InternalServerError from '../error/InternalServerError';
import databaseSchema from './databaseSchema';
import { connectToDatabase, defineDatabaseModels } from '../database/databaseHelper';
import { Sequelize as Seq } from 'sequelize';

export async function setupDatabaseConnection(): Promise<{ database: Seq; models: any[] }> {
  const databaseName = getVariable('DATABASE_NAME');
  const username = getVariable('DATABASE_USERNAME');
  const password = getVariable('DATABASE_PASSWORD');
  const host = getVariable('DATABASE_HOST');

  const database = await connectToDatabase({
    database: databaseName,
    username,
    password,
    options: {
      host,
      dialect: 'postgres',
      operatorsAliases: false,
      pool: {
        max: 1
      }
    }
  });

  if (!database) {
    throw new InternalServerError('Unable to create database connection');
  }

  const models = await defineDatabaseModels(database, databaseSchema);

  return {
    database,
    models
  };
}
