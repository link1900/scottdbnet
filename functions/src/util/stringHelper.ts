import { isArray, isString, isEmpty, isObject, isNil } from 'lodash';
import zlib from 'zlib';
import crypto from 'crypto';

export { isString };

export function arrayToString(fields: Array<string | null | undefined> = [], separator: string = ' '): string {
  if (!isArray(fields)) {
    return '';
  }
  return fields
    .filter(x => isString(x))
    .map(field => field && field.trim())
    .filter(x => !isEmpty(x))
    .join(separator)
    .trim();
}

export function filterForOnlyLetters(value?: string, regex?: string | RegExp): string {
  if (!value) {
    return '';
  }

  return value.replace(/[^a-zA-Z\s]/g, '');
}

export function objectToString(object: any): string {
  if (!isObject(object)) {
    return '';
  }

  return JSON.stringify(object);
}

export function stringToObject(string: string): any {
  try {
    const result = JSON.parse(string);
    if (!result) {
      return null;
    }
    return result;
  } catch (error) {
    return null;
  }
}

export function base64Encode(text?: string): string {
  if (!isString(text)) {
    return '';
  }
  return Buffer.from(text, 'ascii').toString('base64');
}

export function base64Decode(text?: string): string {
  if (!isString(text)) {
    return '';
  }
  return Buffer.from(text, 'base64').toString('ascii');
}

export function getHashForString(string: string): string {
  return crypto
    .createHash('sha1')
    .update(string)
    .digest('base64');
}

export function getHash(something: any): string {
  return getHashForString(anyToString(something));
}

export function anyToString(value: any): string {
  try {
    return JSON.stringify({
      value
    });
  } catch (error) {
    return '';
  }
}

export function stringToAny(string: string): any {
  try {
    const result = JSON.parse(string);
    if (result === undefined || result === null || result.value === undefined) {
      return undefined;
    }
    return result.value;
  } catch (error) {
    return undefined;
  }
}

export async function zipStringToString(string: string): Promise<string> {
  return new Promise((resolve, reject) => {
    zlib.gzip(string, (err, buffer: Buffer) => {
      if (err) {
        return reject(err);
      }
      return resolve(buffer.toString('base64'));
    });
  });
}

export async function unzipStringToString(zipString: string): Promise<string> {
  return new Promise((resolve, reject) => {
    zlib.gunzip(new Buffer(zipString, 'base64'), (err, buffer) => {
      if (err) {
        return reject(err);
      }
      return resolve(buffer.toString());
    });
  });
}

export async function serializeToString(something: any): Promise<string> {
  return zipStringToString(anyToString(something));
}

export async function deserializeFromString(string: string): Promise<any> {
  if (!isString(string)) {
    return undefined;
  }
  const unzippedString = await unzipStringToString(string);
  return stringToAny(unzippedString);
}
