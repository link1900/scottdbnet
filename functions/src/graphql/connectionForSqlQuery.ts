import { base64Decode, base64Encode, objectToString, stringToObject } from '../util/stringHelper';
import InternalServerError from '../error/InternalServerError';
import { Connection, ConnectionArguments } from './graphqlSchemaTypes';
import { SelectQueryBuilder } from 'typeorm';
import { isNil, reverse, get, set } from 'lodash';
import UserInputError from '../error/UserInputError';
import { InvalidFieldReason } from '../error/InvalidFieldReason';
import { BaseModel } from '../database/BaseModel';
import { unwrap } from '../util/objectHelper';

function encodeCursor(value?: { [key: string]: any }): string {
  if (!value) {
    throw new InternalServerError('Failed trying to encode cursor for graphql relay connection');
  }
  return base64Encode(objectToString(value));
}

function decodeCursor(cursor?: string): { [key: string]: any } {
  if (!cursor) {
    throw new InternalServerError('Invalid cursor provided for graphql relay connection');
  }
  return stringToObject(base64Decode(cursor));
}

const emptyConnection = {
  pageInfo: {
    hasNextPage: false,
    hasPreviousPage: false
  },
  edges: []
};

type QueryOrder = 'ASC' | 'DESC';

function getSortOrderForQueryBuilder<T>(queryBuilder: SelectQueryBuilder<T>): QueryOrder {
  const orderByKeys = Object.keys(queryBuilder.expressionMap.orderBys);
  const order = queryBuilder.expressionMap.orderBys[orderByKeys[0]];
  if (order === 'ASC') {
    return 'ASC';
  }
  return 'DESC';
}

function verifyAllOrderBysAreSame<T>(queryBuilder: SelectQueryBuilder<T>): boolean {
  const firstSort = getSortOrderForQueryBuilder(queryBuilder);
  const orderByKeys = Object.keys(queryBuilder.expressionMap.orderBys);
  const result = orderByKeys.every(key => queryBuilder.expressionMap.orderBys[key] === firstSort);
  if (!result) {
    throw new InternalServerError('Cannot page query as all order bys must have a consistent sort direction');
  }
  return result;
}

function buildCursorForQueryAndNode<T>(queryBuilder: SelectQueryBuilder<T>, node: any): { [key: string]: any } {
  const orderByKeys = Object.keys(queryBuilder.expressionMap.orderBys);
  const name = getNameForQueryBuilder(queryBuilder);
  return orderByKeys.reduce((ob, key) => {
    const cleanKey = key.replace(`${name}.`, '');
    return {
      ...ob,
      [cleanKey]: get(node, cleanKey)
    };
  }, {});
}

function flipOrder(order: QueryOrder): QueryOrder {
  if (order === 'ASC') {
    return 'DESC';
  }

  return 'ASC';
}

function flippedOrderBy<T>(queryBuilder: SelectQueryBuilder<T>): { [key: string]: any } {
  const firstSort = getSortOrderForQueryBuilder(queryBuilder);
  const flippedSort = flipOrder(firstSort);
  const orderByKeys = Object.keys(queryBuilder.expressionMap.orderBys);
  return orderByKeys.reduce((ob, key) => {
    set(ob, key, flippedSort);
    return ob;
  }, {});
}

function getNameForQueryBuilder<T>(queryBuilder: SelectQueryBuilder<T>): string {
  if (!queryBuilder.expressionMap.mainAlias) {
    return '';
  }
  return queryBuilder.expressionMap.mainAlias.name;
}

export async function runQueryBuilderAsConnection<T extends BaseModel>(
  queryBuilder: SelectQueryBuilder<T>,
  args: ConnectionArguments
): Promise<Connection<T>> {
  let limit: number = 10;
  const name = getNameForQueryBuilder(queryBuilder);
  const hasFirst = !isNil(args.first);
  const hasLast = !isNil(args.last);
  const hasBefore = !isNil(args.before);
  const hasAfter = !isNil(args.after);
  const isReversed = (hasBefore || hasLast) && !hasFirst;

  if (hasFirst && (typeof args.first !== 'number' || args.first < 0)) {
    throw new UserInputError('args.first must be a positive integer', 'args.first', InvalidFieldReason.INVALID);
  }

  if (hasLast && (typeof args.last !== 'number' || args.last < 0)) {
    throw new UserInputError('args.last must be a positive integer', 'args.last', InvalidFieldReason.INVALID);
  }

  if (hasLast && hasFirst) {
    throw new UserInputError('Cannot provide both first and last arguments.', 'args', InvalidFieldReason.REQUIRED);
  }

  // get sortDirection from first order by or default to asc
  const sortDirection: QueryOrder = getSortOrderForQueryBuilder(queryBuilder);

  // must always order by id field
  const orderBys = Object.keys(queryBuilder.expressionMap.orderBys);
  const foundIdOrder = orderBys.find(orderBy => orderBy === 'id');
  if (!foundIdOrder) {
    queryBuilder.addOrderBy('id', sortDirection);
  }

  verifyAllOrderBysAreSame(queryBuilder);

  if (hasAfter) {
    const cursor = decodeCursor(args.after);
    const operator = sortDirection === 'ASC' ? '>' : '<';
    const columnNames = Object.keys(cursor);
    const sortColumns = columnNames.map(columnName => `${name}.${columnName}`);
    const valueParamNames = columnNames.map(columnName => `:${columnName}`);
    queryBuilder.andWhere(`(${sortColumns.join(', ')}) ${operator} (${valueParamNames.join(', ')})`, cursor);
  }

  if (hasBefore) {
    const cursor = decodeCursor(args.before);
    const operator = sortDirection === 'DESC' ? '>' : '<';
    const columnNames = Object.keys(cursor);
    const sortColumns = columnNames.map(columnName => `${name}.${columnName}`);
    const valueParamNames = columnNames.map(columnName => `:${columnName}`);
    queryBuilder.andWhere(`(${sortColumns.join(', ')}) ${operator} (${valueParamNames.join(', ')})`, cursor);
  }

  // work out the correct limit size from the args
  if (hasFirst) {
    limit = unwrap(args.first);
  }

  if (hasLast) {
    limit = unwrap(args.last);
  }

  if (limit === 0) {
    return emptyConnection;
  }

  // work out the sort
  if (isReversed) {
    queryBuilder.orderBy(flippedOrderBy(queryBuilder));
  }

  // apply limit
  queryBuilder.take(limit + 1);

  // run the constructed query
  // we are fetching one extra edge to see if there are more results
  let entities = await queryBuilder.getMany();
  if (entities.length === 0) {
    return emptyConnection;
  }

  const hasMore = limit ? entities.length > limit : false;
  if (hasMore) {
    // this is to remove the entity that was over-fetched
    entities.pop();
  }

  // flip results if we are paging backwards
  if (isReversed) {
    entities = reverse(entities);
  }

  // convert results to relay edges and nodes
  const edges = entities.map(entity => ({
    cursor: encodeCursor(buildCursorForQueryAndNode(queryBuilder, entity)),
    node: entity
  }));

  return {
    pageInfo: {
      hasNextPage: hasMore && hasFirst ? true : false,
      hasPreviousPage: hasMore && hasLast ? true : false
    },
    edges
  };
}
