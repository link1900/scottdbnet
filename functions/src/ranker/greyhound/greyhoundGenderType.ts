import { gql } from 'apollo-server-express';
import { createEnum } from '../../graphql/graphqlSchemaBuilders';

export const greyhoundGenderDefinition = gql`
  enum GreyhoundGender {
    DOG
    BITCH
  }
`;

export enum GreyhoundGender {
  DOG = 'dog',
  BITCH = 'bitch'
}

export const greyhoundGenderResolver = {
  DOG: GreyhoundGender.DOG,
  BITCH: GreyhoundGender.BITCH
};

export const greyhoundGenderType = createEnum({
  name: 'GreyhoundGender',
  definition: greyhoundGenderDefinition,
  resolver: greyhoundGenderResolver
});
