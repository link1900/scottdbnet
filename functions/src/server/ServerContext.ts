import { Context } from '../graphql/graphqlSchemaTypes';
import { DataLoaders } from './dataLoaders';

export default class ServerContext implements Context {
  loaders: DataLoaders;

  constructor(loaders: DataLoaders) {
    this.loaders = loaders;
  }
}
