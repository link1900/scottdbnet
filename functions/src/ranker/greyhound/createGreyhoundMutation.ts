import { gql } from 'apollo-server-core';
import { createMutation } from '../../graphql/graphqlSchemaBuilders';
import ServerContext from '../../server/ServerContext';
import { Greyhound } from './Greyhound';
import { MutationInput, MutationPayload } from '../../graphql/graphqlSchemaTypes';
import UserInputError from '../../error/UserInputError';
import { InvalidFieldReason } from '../../error/InvalidFieldReason';
import { Role } from '../../server/Role';
import { doesGreyhoundNameExist } from './greyhoundHelper';

export const createGreyhoundMutationDefinition = gql`
  input CreateGreyhoundInput {
    name: TextField! @notEmpty @upperCase
    color: TextField
    gender: GreyhoundGender
    dateOfBirth: DateTime
    clientMutationId: String
  }

  type CreateGreyhoundPayload {
    greyhound: Greyhound
    clientMutationId: String
  }

  extend type Mutation {
    createGreyhound(input: CreateGreyhoundInput!): CreateGreyhoundPayload
  }
`;

export type CreateGreyhoundInput = Partial<Greyhound> & MutationInput;
export type CreateGreyhoundPayload = {
  greyhound: Greyhound;
} & MutationPayload;

export async function createGreyhoundResolver(
  input: CreateGreyhoundInput,
  context: ServerContext
): Promise<CreateGreyhoundPayload> {
  context.checkForRole(Role.ADMIN);
  const greyhoundInput = new Greyhound(input);
  const nameAlreadyExists = await doesGreyhoundNameExist(context, greyhoundInput.name);
  if (nameAlreadyExists) {
    throw new UserInputError(
      `greyhound with name ${greyhoundInput.name} already exists`,
      'name',
      InvalidFieldReason.INVALID
    );
  }
  const greyhound = await context.loaders.greyhound.create(greyhoundInput);
  return {
    greyhound
  };
}

export const createGreyhoundMutation = createMutation({
  name: 'createGreyhound',
  definition: createGreyhoundMutationDefinition,
  resolver: createGreyhoundResolver
});
