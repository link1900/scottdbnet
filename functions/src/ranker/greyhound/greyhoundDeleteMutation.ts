import { gql } from 'apollo-server-core';
import { createMutation } from '../../graphql/graphqlSchemaBuilders';
import ServerContext from '../../server/ServerContext';
import { MutationInput, MutationPayload } from '../../graphql/graphqlSchemaTypes';
import { Role } from '../../server/Role';
import NotFoundError from '../../error/NotFoundError';
import InternalServerError from '../../error/InternalServerError';
import UserInputError from '../../error/UserInputError';
import { InvalidFieldReason } from '../../error/InvalidFieldReason';

export const deleteGreyhoundMutationDefinition = gql`
  input DeleteGreyhoundInput {
    id: NodeId!
    clientMutationId: String
  }

  type DeleteGreyhoundPayload {
    deletedId: ID
    clientMutationId: String
  }

  extend type Mutation {
    deleteGreyhound(input: DeleteGreyhoundInput!): DeleteGreyhoundPayload
  }
`;

export type DeleteGreyhoundInput = { id?: string } & MutationInput;
export type DeleteGreyhoundPayload = {
  deletedId: string;
} & MutationPayload;

export async function deleteGreyhoundResolver(
  input: DeleteGreyhoundInput,
  context: ServerContext
): Promise<DeleteGreyhoundPayload> {
  context.checkForRole(Role.ADMIN);
  const { id } = input;
  if (!id) {
    throw new UserInputError(`id is required`, 'id', InvalidFieldReason.REQUIRED);
  }
  const greyhound = await context.loaders.greyhound.findById(id);
  if (!greyhound) {
    throw new NotFoundError(`greyhound with id ${id} does not exist`);
  }
  const deletedId = greyhound.nodeId;
  const result = await context.loaders.greyhound.delete(greyhound);
  if (result === false) {
    throw new InternalServerError(`Failed to delete greyhound ${id}`);
  }
  return {
    deletedId
  };
}

export const deleteGreyhoundMutation = createMutation({
  name: 'deleteGreyhound',
  definition: deleteGreyhoundMutationDefinition,
  resolver: deleteGreyhoundResolver
});
