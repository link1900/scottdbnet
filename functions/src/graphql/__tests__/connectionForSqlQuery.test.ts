import { closeDatabaseConnection } from '../../server/serverHelper';
import { Greyhound } from '../../ranker/greyhound/Greyhound';
import { runQueryBuilderAsConnection } from '../connectionForSqlQuery';
import ServerContext from '../../server/ServerContext';
import { base64Encode, objectToString } from '../../util/stringHelper';
import { getContextAndFixture } from '../../server/__tests__/testHelpers';

describe('connectionForSqlQuery', () => {
  let context: ServerContext;
  let greyhounds: Greyhound[];

  beforeEach(async () => {
    const { context: someContext } = await getContextAndFixture();
    context = someContext;
    await context.loaders.greyhound.repository.clear();
    greyhounds = await Promise.all(
      Array(10)
        .fill(0)
        .map((i, index) =>
          context.loaders.greyhound.create(new Greyhound({ name: `${String.fromCharCode(index + 65)}` }))
        )
    );
  });

  afterAll(async () => {
    await closeDatabaseConnection();
  });

  describe('#runQueryBuilderAsConnection', () => {
    it('returns the 5 edges when requested', async () => {
      const queryDefinition = context.loaders.greyhound.repository.createQueryBuilder('greyhound');
      const args = { first: 5 };
      const { edges, pageInfo } = await runQueryBuilderAsConnection(queryDefinition, args);
      expect(edges.length).toEqual(5);
      expect(pageInfo.hasNextPage).toEqual(true);
      expect(pageInfo.hasPreviousPage).toEqual(false);
    });

    it('returns an empty connection if no data is returned by query', async () => {
      await context.loaders.greyhound.repository.clear();
      const queryDefinition = context.loaders.greyhound.repository.createQueryBuilder('greyhound');
      const args = { first: 5 };
      const { edges, pageInfo } = await runQueryBuilderAsConnection(queryDefinition, args);
      expect(edges.length).toEqual(0);
      expect(pageInfo.hasNextPage).toEqual(false);
      expect(pageInfo.hasPreviousPage).toEqual(false);
    });

    it('returns default max 10 results when passed no parameters', async () => {
      const queryDefinition = context.loaders.greyhound.repository.createQueryBuilder('greyhound');
      await Promise.all(
        Array(2)
          .fill(0)
          .map((i, index) =>
            context.loaders.greyhound.create(new Greyhound({ name: `${String.fromCharCode(index + 75)}` }))
          )
      );
      const args = { first: 10 };
      const { edges, pageInfo } = await runQueryBuilderAsConnection(queryDefinition, args);
      expect(edges.length).toEqual(10);
      expect(pageInfo.hasNextPage).toEqual(true);
      expect(pageInfo.hasPreviousPage).toEqual(false);
    });

    describe('Check various return lengths', () => {
      [3, 5, 7, 10].forEach(firstArg => {
        it(`returns the ${firstArg} edges when requested`, async () => {
          const queryDefinition = context.loaders.greyhound.repository.createQueryBuilder('greyhound');
          const args = { first: firstArg };
          const { edges, pageInfo } = await runQueryBuilderAsConnection(queryDefinition, args);
          expect(edges.length).toEqual(firstArg);
          expect(pageInfo.hasNextPage).toEqual(firstArg === 10 ? false : true);
          expect(pageInfo.hasPreviousPage).toEqual(false);
        });
      });
    });

    it('applies after argument correctly', async () => {
      const greyhoundH = greyhounds.find(g => g.name === 'H');
      if (!greyhoundH) {
        throw new Error('test data issue');
      }
      const cursor = base64Encode(objectToString({ name: greyhoundH.name, id: greyhoundH.id }));
      const queryDefinition = context.loaders.greyhound.repository
        .createQueryBuilder('greyhound')
        .orderBy('name', 'ASC');
      const args = { after: cursor };
      const { edges, pageInfo } = await runQueryBuilderAsConnection(queryDefinition, args);
      expect(edges.length).toEqual(2);
      expect(edges[0].node.name).toEqual('I');
      expect(edges[1].node.name).toEqual('J');
      expect(pageInfo.hasNextPage).toEqual(false);
      expect(pageInfo.hasPreviousPage).toEqual(false);
    });

    it('applies before argument correctly', async () => {
      const greyhoundH = greyhounds.find(g => g.name === 'H');
      if (!greyhoundH) {
        throw new Error('test data issue');
      }
      const cursor = base64Encode(objectToString({ name: greyhoundH.name, id: greyhoundH.id }));
      const queryDefinition = context.loaders.greyhound.repository
        .createQueryBuilder('greyhound')
        .orderBy('name', 'ASC');
      const args = { before: cursor };
      const { edges, pageInfo } = await runQueryBuilderAsConnection(queryDefinition, args);
      expect(edges.length).toEqual(7);
      expect(edges[0].node.name).toEqual('A');
      expect(edges[1].node.name).toEqual('B');
      expect(edges[2].node.name).toEqual('C');
      expect(edges[3].node.name).toEqual('D');
      expect(edges[4].node.name).toEqual('E');
      expect(edges[5].node.name).toEqual('F');
      expect(edges[6].node.name).toEqual('G');
      expect(pageInfo.hasNextPage).toEqual(false);
      expect(pageInfo.hasPreviousPage).toEqual(false);
    });

    it('applies first argument correctly', async () => {
      const queryDefinition = context.loaders.greyhound.repository
        .createQueryBuilder('greyhound')
        .orderBy('name', 'ASC');
      const args = { first: 5 };
      const { edges, pageInfo } = await runQueryBuilderAsConnection(queryDefinition, args);
      expect(edges.length).toEqual(5);
      expect(edges[0].node.name).toEqual('A');
      expect(edges[1].node.name).toEqual('B');
      expect(edges[2].node.name).toEqual('C');
      expect(edges[3].node.name).toEqual('D');
      expect(edges[4].node.name).toEqual('E');
      expect(pageInfo.hasNextPage).toEqual(true);
      expect(pageInfo.hasPreviousPage).toEqual(false);
    });

    it('applies last argument correctly', async () => {
      const queryDefinition = context.loaders.greyhound.repository
        .createQueryBuilder('greyhound')
        .orderBy('name', 'ASC');
      const args = { last: 5 };
      const { edges, pageInfo } = await runQueryBuilderAsConnection(queryDefinition, args);
      expect(edges.length).toEqual(5);
      expect(edges[0].node.name).toEqual('F');
      expect(edges[1].node.name).toEqual('G');
      expect(edges[2].node.name).toEqual('H');
      expect(edges[3].node.name).toEqual('I');
      expect(edges[4].node.name).toEqual('J');
      expect(pageInfo.hasNextPage).toEqual(false);
      expect(pageInfo.hasPreviousPage).toEqual(true);
    });

    it('applies after and last arguments correctly', async () => {
      const queryDefinition = context.loaders.greyhound.repository
        .createQueryBuilder('greyhound')
        .orderBy({ name: 'ASC' });
      const args = { first: 2 };
      const { edges, pageInfo } = await runQueryBuilderAsConnection<Greyhound>(queryDefinition, args);
      expect(edges.length).toEqual(2);
      expect(pageInfo.hasNextPage).toEqual(true);
      expect(pageInfo.hasPreviousPage).toEqual(false);
      expect(edges[0].node.name).toEqual('A');
      expect(edges[1].node.name).toEqual('B');

      // check the ordering continues for page 2
      const queryDefinition2nd = context.loaders.greyhound.repository
        .createQueryBuilder('greyhound')
        .orderBy({ name: 'ASC' });
      const page2Args = { last: 2, after: edges[1].cursor };
      const { edges: page2Edges, pageInfo: page2PageInfo } = await runQueryBuilderAsConnection<Greyhound>(
        queryDefinition2nd,
        page2Args
      );
      expect(page2Edges.length).toEqual(2);
      expect(page2PageInfo.hasNextPage).toEqual(false);
      expect(page2PageInfo.hasPreviousPage).toEqual(true);
      expect(page2Edges[0].node.name).toEqual('I');
      expect(page2Edges[1].node.name).toEqual('J');
    });

    it('applies before and first arguments correctly', async () => {
      const queryDefinition = context.loaders.greyhound.repository
        .createQueryBuilder('greyhound')
        .orderBy({ name: 'ASC' });
      const args = { last: 2 };
      const { edges, pageInfo } = await runQueryBuilderAsConnection<Greyhound>(queryDefinition, args);
      expect(edges.length).toEqual(2);
      expect(pageInfo.hasNextPage).toEqual(false);
      expect(pageInfo.hasPreviousPage).toEqual(true);
      expect(edges[0].node.name).toEqual('I');
      expect(edges[1].node.name).toEqual('J');

      // check the ordering continues for page 2
      const queryDefinition2nd = context.loaders.greyhound.repository
        .createQueryBuilder('greyhound')
        .orderBy({ name: 'ASC' });
      const page2Args = { first: 2, before: edges[1].cursor };
      const { edges: page2Edges, pageInfo: page2PageInfo } = await runQueryBuilderAsConnection<Greyhound>(
        queryDefinition2nd,
        page2Args
      );
      expect(page2Edges.length).toEqual(2);
      expect(page2PageInfo.hasNextPage).toEqual(true);
      expect(page2PageInfo.hasPreviousPage).toEqual(false);
      expect(page2Edges[0].node.name).toEqual('A');
      expect(page2Edges[1].node.name).toEqual('B');
    });

    it('applies after and first arguments correctly', async () => {
      const queryDefinition = context.loaders.greyhound.repository
        .createQueryBuilder('greyhound')
        .orderBy({ name: 'DESC' });
      const args = { first: 5 };
      const { edges, pageInfo } = await runQueryBuilderAsConnection<Greyhound>(queryDefinition, args);
      expect(edges.length).toEqual(5);
      expect(pageInfo.hasNextPage).toEqual(true);
      expect(pageInfo.hasPreviousPage).toEqual(false);
      expect(edges[0].node.name).toEqual('J');
      expect(edges[1].node.name).toEqual('I');
      expect(edges[2].node.name).toEqual('H');
      expect(edges[3].node.name).toEqual('G');
      expect(edges[4].node.name).toEqual('F');

      // check the ordering continues for page 2
      const page2Args = { first: 3, after: edges[4].cursor };
      const { edges: page2Edges, pageInfo: page2PageInfo } = await runQueryBuilderAsConnection<Greyhound>(
        queryDefinition,
        page2Args
      );
      expect(page2Edges.length).toEqual(3);
      expect(page2PageInfo.hasNextPage).toEqual(true);
      expect(page2PageInfo.hasPreviousPage).toEqual(false);
      expect(page2Edges[0].node.name).toEqual('E');
      expect(page2Edges[1].node.name).toEqual('D');
      expect(page2Edges[2].node.name).toEqual('C');
    });

    it('applies and before and last arguments correctly', async () => {
      const queryDefinition = context.loaders.greyhound.repository
        .createQueryBuilder('greyhound')
        .orderBy({ name: 'ASC' });
      const args = { last: 5 };
      const { edges, pageInfo } = await runQueryBuilderAsConnection<Greyhound>(queryDefinition, args);
      expect(edges.length).toEqual(5);
      expect(pageInfo.hasNextPage).toEqual(false);
      expect(pageInfo.hasPreviousPage).toEqual(true);
      expect(edges[0].node.name).toEqual('F');
      expect(edges[1].node.name).toEqual('G');
      expect(edges[2].node.name).toEqual('H');
      expect(edges[3].node.name).toEqual('I');
      expect(edges[4].node.name).toEqual('J');

      // check the ordering continues for page 2
      const queryDefinition2nd = context.loaders.greyhound.repository
        .createQueryBuilder('greyhound')
        .orderBy({ name: 'ASC' });
      const page2Args = { last: 2, before: edges[0].cursor };
      const { edges: page2Edges, pageInfo: page2PageInfo } = await runQueryBuilderAsConnection<Greyhound>(
        queryDefinition2nd,
        page2Args
      );
      expect(page2Edges.length).toEqual(2);
      expect(page2PageInfo.hasNextPage).toEqual(false);
      expect(page2PageInfo.hasPreviousPage).toEqual(true);
      expect(page2Edges[0].node.name).toEqual('D');
      expect(page2Edges[1].node.name).toEqual('E');
    });

    it('throws an error when first is negative', async () => {
      const queryDefinition = context.loaders.greyhound.repository
        .createQueryBuilder('greyhound')
        .orderBy({ name: 'ASC' });
      const args = { first: -5 };

      try {
        await runQueryBuilderAsConnection<Greyhound>(queryDefinition, args);
        expect(true).toBeFalsy();
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });

    it('throws an error when last is negative', async () => {
      const queryDefinition = context.loaders.greyhound.repository
        .createQueryBuilder('greyhound')
        .orderBy({ name: 'ASC' });
      const args = { last: -5 };

      try {
        await runQueryBuilderAsConnection<Greyhound>(queryDefinition, args);
        expect(true).toBeFalsy();
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });

    it('throws an error when first and last is provided', async () => {
      const queryDefinition = context.loaders.greyhound.repository
        .createQueryBuilder('greyhound')
        .orderBy({ name: 'ASC' });
      const args = { first: 5, last: 5 };

      try {
        await runQueryBuilderAsConnection<Greyhound>(queryDefinition, args);
        expect(true).toBeFalsy();
      } catch (e) {
        expect(e).toMatchSnapshot();
      }
    });
  });
});
