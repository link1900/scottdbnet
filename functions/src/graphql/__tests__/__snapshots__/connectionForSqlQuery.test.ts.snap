// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connectionForSqlQuery #runQueryBuilderAsConnection throws an error when first and last is provided 1`] = `[Error: Cannot provide both first and last arguments.]`;

exports[`connectionForSqlQuery #runQueryBuilderAsConnection throws an error when first is negative 1`] = `[Error: args.first must be a positive integer]`;

exports[`connectionForSqlQuery #runQueryBuilderAsConnection throws an error when last is negative 1`] = `[Error: args.last must be a positive integer]`;
